src/librustc_mir/hair/pattern/mod.rs

```diff
@@ -974,10 +974,25 @@ impl<'a, 'tcx> PatternContext<'a, 'tcx> {
                 PatternKind::Wild
             }
             ty::Adt(adt_def, _) if !self.tcx.has_attr(adt_def.did, "structural_match") => {
-                let msg = format!("to use a constant of type `{}` in a pattern, \
-                                    `{}` must be annotated with `#[derive(PartialEq, Eq)]`",
-                                    self.tcx.def_path_str(adt_def.did),
-                                    self.tcx.def_path_str(adt_def.did));
+                let path = self.tcx.def_path_str(adt_def.did);
+                let msg = format!(
+                    "to use a constant of type `{}` in a pattern, \
+                     `{}` must be annotated with `#[derive(PartialEq, Eq)]`",
+                    path,
+                    path,
+                );
+                self.tcx.sess.span_err(span, &msg);
+                PatternKind::Wild
+            }
+            ty::Ref(_, ty::TyS { sty: ty::Adt(adt_def, _), .. }, _)
```

Comment:
This just papers over the reference case. I think a double reference will still trigger everything that was triggered without this arm

Reply:
But is this handling correct for this case? Would it be reasonable to add a loop dereferencing indefinite amount of times to determine this case? (I assumed that because of match ergonomics and this being a pattern multiple borrows would have been already handled.)

Reply:
just for the error case that would be ok, but add a `HACK` comment making it clear that anything but erroring and using `PatternKind::Wild` is wrong here.